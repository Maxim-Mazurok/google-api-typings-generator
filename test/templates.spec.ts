import {readFileSync} from 'node:fs';
import path, {join} from 'node:path';
import doT from 'dot';
import dedent from 'dedent';
import {fileURLToPath} from 'node:url';
import {TemplateData} from '../src/template/index.js';
const __dirname = path.dirname(fileURLToPath(import.meta.url));

beforeAll(() => {
  doT.templateSettings.strip = false;
});

describe('readme', () => {
  it('works', () => {
    const template = readFileSync(
      join(__dirname, '..', 'src', 'template', 'readme.dot'),
      'utf-8'
    );
    const data: TemplateData = {
      restDescription: {
        name: 'my-name',
        title: 'My Types',
        version: 'v1beta',
        documentationLink: 'http://docs.io',
        auth: {
          oauth2: {
            scopes: {
              'http://my.scope': {description: 'my scope'},
              'http://your.scope': {description: 'your scope'},
            },
          },
        },
        resources: {
          firstResource: {
            methods: {
              firstMethod: {
                description: 'Method Number One',
                httpMethod: 'GET',
                path: 'some/path/1',
                id: 'thirdNamespace.firstMethod',
              },
            },
          },
        },
      },
      restDescriptionSource: 'http://x.com',
      namespaces: ['admin', 'directory'], // TODO: need to actually add this to template data I guess
      majorAndMinorVersion: '0.0',
      packageName: 'my-name-v1beta',
    };

    const result = doT.compile(template)(data);

    expect(result).toMatchSnapshot();
  });

  it.skip('use me for small tests during development', () => {
    const template = dedent`
      gapi.client.load('{{=it.url}}', () => {
        // now we can use:{{~ it.namespaces :namespace }}
        // gapi.client.{{=namespace}}{{~}}
      });
    `;
    const data = {
      url: 'http://x.com',
      namespaces: ['admin', 'directory'],
    };
    const result = doT.compile(template)(data);

    const expected = dedent`
      gapi.client.load('http://x.com', () => {
        // now we can use:
        // gapi.client.admin
        // gapi.client.directory
      });
    `;
    expect(result).toBe(expected);
  });
});
