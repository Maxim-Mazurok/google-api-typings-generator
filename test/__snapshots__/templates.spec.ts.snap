// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`readme > works 1`] = `
"# TypeScript typings for My Types v1beta

undefined
For detailed description please check [documentation](http://docs.io).

## Installing

Install typings for My Types:

\`\`\`
npm install @types/my-name-v1beta --save-dev
\`\`\`

## Usage

You need to initialize Google API client in your code:

\`\`\`typescript
gapi.load('client', () => {
  // now we can use gapi.client
  // ...
});
\`\`\`

Then load api client wrapper:

\`\`\`typescript
gapi.client.load('http://x.com', () => {
  // now we can use:
  // gapi.client.admin
  // gapi.client.directory
});
\`\`\`

\`\`\`typescript
// Deprecated, use discovery document URL, see https://github.com/google/google-api-javascript-client/blob/master/docs/reference.md#----gapiclientloadname----version----callback--
gapi.client.load('my-name', 'v1beta', () => {
  // now we can use:
  // gapi.client.admin
  // gapi.client.directory
});
\`\`\`

Don't forget to authenticate your client before sending any request to resources:

\`\`\`typescript
// declare client_id registered in Google Developers Console
var client_id = '',
  scope = [
      // my scope
      'http://my.scope',

      // your scope
      'http://your.scope',
    ],
    immediate = true;
// ...

gapi.auth.authorize(
  { client_id: client_id, scope: scope, immediate: immediate },
  authResult => {
    if (authResult && !authResult.error) {
        /* handle successful authorization */
    } else {
        /* handle authorization error */
    }
});
\`\`\`

After that you can use My Types resources: <!-- TODO: make this work for multiple namespaces -->

\`\`\`typescript

/*
Method Number One
*/
await gapi.client.thirdNamespace.firstMethod({  });
\`\`\`
"
`;
